# Assignment: Data Wrangling

In this assignment, we will practice data wrangling techniques on real world data.

## NYC film permits

New York City is a very popular film/tv shooting location. In order to assert the exclusive usage of city properties (e.g. sidewalk, park, etc.) for this purpose, a permit is required. New York Mayor's Office of Media and Entertainment release all granted [film permits data](https://data.cityofnewyork.us/City-Government/Film-Permits/tg4x-b46p) to the public. In this part, we will do some data wrangling on this data.

Let us first retrieve the data!

In [5]:
import pandas as pd

url = "https://data.cityofnewyork.us/api/views/tg4x-b46p/rows.csv"
    
film_permits = pd.read_csv(url)

Your task for this part is to clean up this data. Specifically, the cleaned up dataframe should have the following properties:
1. All columns must have the correct data type.  In particular, `StartDateTime`, `EndDateTime`, `EnteredOn` should have `datetime64` as data type.
2. The `ZipCode(s)` column may contain multiple zip codes in one cell. In the cleaned up dataframe, we will replace the `ZipCode(s)` column with `ZipCode` column, where each cell in the `ZipCode` column only contains a single zip code. Rows with multiple zip codes in the input data frame should be repeated for each zip code in the output dataframe.
3. The output dataframe should not contain any missing data.

In [6]:
film_permits.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7870 entries, 0 to 7869
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   EventID            7870 non-null   int64 
 1   EventType          7870 non-null   object
 2   StartDateTime      7870 non-null   object
 3   EndDateTime        7870 non-null   object
 4   EnteredOn          7870 non-null   object
 5   EventAgency        7870 non-null   object
 6   ParkingHeld        7870 non-null   object
 7   Borough            7870 non-null   object
 8   CommunityBoard(s)  7868 non-null   object
 9   PolicePrecinct(s)  7868 non-null   object
 10  Category           7870 non-null   object
 11  SubCategoryName    7870 non-null   object
 12  Country            7870 non-null   object
 13  ZipCode(s)         7868 non-null   object
dtypes: int64(1), object(13)
memory usage: 860.9+ KB


In [7]:
#film_permits orignally has 7870 rows
film_permits = film_permits.dropna()
film_permits #now has 7868 rows

EventID                      EventType           StartDateTime  \
0      696255                Shooting Permit  02/17/2023 09:00:00 AM   
1      714139                Shooting Permit  05/12/2023 01:00:00 PM   
2      705334                Shooting Permit  04/10/2023 09:00:00 AM   
3      746696                Shooting Permit  11/07/2023 06:00:00 AM   
4      717328  Theater Load in and Load Outs  05/31/2023 12:01:00 AM   
...       ...                            ...                     ...   
7865   793722                Shooting Permit  06/19/2024 12:00:00 PM   
7866   793926                Shooting Permit  06/20/2024 09:00:00 AM   
7867   793783  Theater Load in and Load Outs  06/20/2024 12:01:00 AM   
7868   785081  Theater Load in and Load Outs  06/20/2024 12:01:00 AM   
7869   792643                Shooting Permit  06/20/2024 07:00:00 AM   

                 EndDateTime               EnteredOn  \
0     02/18/2023 12:00:00 PM  02/14/2023 10:47:33 PM   
1     05/13/2023 05:00:00 AM  05/04/2023 02:27:51 PM   
2     04/10/2023 10:00:00 PM  03/30/2023 05:17:29 PM   
3     11/07/2023 10:00:00 PM  10/31/2023 12:08:00 PM   
4     06/01/2023 06:00:00 AM  05/17/2023 11:55:51 AM   
...                      ...                     ...   
7865  06/20/2024 04:00:00 AM  06/17/2024 11:01:29 AM   
7866  06/20/2024 10:00:00 PM  06/17/2024 10:24:41 PM   
7867  06/24/2024 11:59:00 PM  06/17/2024 01:21:55 PM   
7868  06/21/2024 06:00:00 AM  05/16/2024 12:54:50 PM   
7869  06/20/2024 10:00:00 PM  06/12/2024 05:10:02 PM   

                                  EventAgency  \
0     Mayor's Office of Media & Entertainment   
1     Mayor's Office of Media & Entertainment   
2     Mayor's Office of Media & Entertainment   
3     Mayor's Office of Media & Entertainment   
4     Mayor's Office of Media & Entertainment   
...                                       ...   
7865  Mayor's Office of Media & Entertainment   
7866  Mayor's Office of Media & Entertainment   
7867  Mayor's Office of Media & Entertainment   
7868  Mayor's Office of Media & Entertainment   
7869  Mayor's Office of Media & Entertainment   

                                            ParkingHeld    Borough  \
0     KINGSLAND AVENUE between DEAD END and GREENPOI...   Brooklyn   
1     WEST   26 STREET between 12 AVENUE and 11 AVEN...  Manhattan   
2     SOUTH STREET between BROAD STREET and OLD SLIP...  Manhattan   
3     NORTH HENRY STREET between GREENPOINT AVENUE a...   Brooklyn   
4     WEST   55 STREET between 11 AVENUE and 12 AVEN...  Manhattan   
...                                                 ...        ...   
7865  BUTLER STREET between 3 AVENUE and 4 AVENUE,  ...   Brooklyn   
7866  SCHERMERHORN STREET between COURT STREET and B...   Brooklyn   
7867  WEST   35 STREET between 8 AVENUE and 9 AVENUE...  Manhattan   
7868  FROST STREET between DEBEVOISE AVENUE and MORG...   Brooklyn   
7869  COLUMBUS AVENUE between WEST   62 STREET and W...  Manhattan   

     CommunityBoard(s) PolicePrecinct(s)           Category  SubCategoryName  \
0                    1                94         Television   Cable-episodic   
1                    4                10         Television  Episodic series   
2                 1, 3              1, 5               Film          Feature   
3                    1                94         Commercial       Commercial   
4                    4                18            Theater          Theater   
...                ...               ...                ...              ...   
7865                 6                78               Film          Feature   
7866           2, 3, 4    79, 81, 83, 84  Still Photography   Not Applicable   
7867                 4                14            Theater          Theater   
7868                 1                94            Theater          Theater   
7869              4, 7            18, 20         Television  Episodic series   

                       Country                  ZipCode(s)  
0     United Sta

In [8]:
film_permits.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 7868 entries, 0 to 7869
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   EventID            7868 non-null   int64 
 1   EventType          7868 non-null   object
 2   StartDateTime      7868 non-null   object
 3   EndDateTime        7868 non-null   object
 4   EnteredOn          7868 non-null   object
 5   EventAgency        7868 non-null   object
 6   ParkingHeld        7868 non-null   object
 7   Borough            7868 non-null   object
 8   CommunityBoard(s)  7868 non-null   object
 9   PolicePrecinct(s)  7868 non-null   object
 10  Category           7868 non-null   object
 11  SubCategoryName    7868 non-null   object
 12  Country            7868 non-null   object
 13  ZipCode(s)         7868 non-null   object
dtypes: int64(1), object(13)
memory usage: 922.0+ KB


In [9]:
film_permits.columns = film_permits.columns.str.lower()
film_permits.dtypes

eventid               int64
eventtype            object
startdatetime        object
enddatetime          object
enteredon            object
eventagency          object
parkingheld          object
borough              object
communityboard(s)    object
policeprecinct(s)    object
category             object
subcategoryname      object
country              object
zipcode(s)           object
dtype: object

In [10]:
film_permits.startdatetime = pd.to_datetime(film_permits.startdatetime, format='%m/%d/%Y %I:%M:%S %p')
film_permits.enddatetime = pd.to_datetime(film_permits.enddatetime, format='%m/%d/%Y %I:%M:%S %p')
film_permits.enteredon = pd.to_datetime(film_permits.enteredon, format='%m/%d/%Y %I:%M:%S %p')
film_permits.category = film_permits.category.astype("category")
film_permits.subcategoryname = film_permits.subcategoryname.astype("category")
film_permits.country = film_permits.country.astype("category")
film_permits.borough = film_permits.borough.astype("category")
film_permits.dtypes

eventid                       int64
eventtype                    object
startdatetime        datetime64[ns]
enddatetime          datetime64[ns]
enteredon            datetime64[ns]
eventagency                  object
parkingheld                  object
borough                    category
communityboard(s)            object
policeprecinct(s)            object
category                   category
subcategoryname            category
country                    category
zipcode(s)                   object
dtype: object

In [27]:
film_permits.rename(columns={'zipcode(s)':'zipcode'}, inplace=True)
zipCodes = 13 #the column number for zipcodes
i = 0
while i < len(film_permits):
    film_permits_currRow = film_permits.iloc[[i]]
    zipCodesList = film_permits.iloc[i, zipCodes].split(",")
    if len(zipCodesList) > 1:
        film_permits_top = film_permits.iloc[:i]
        film_permits_bot = film_permits.iloc[i+1:]
    
        for zipcode in zipCodesList:
            film_permits_row_dup = film_permits_currRow
            film_permits_row_dup.iloc[0,zipCodes] = zipcode
            film_permits_top = pd.concat([film_permits_top, film_permits_currRow], axis=0)

        film_permits = pd.concat([film_permits_top, film_permits_bot], axis=0, ignore_index=True)
        
    i +=1
film_permits

eventid                      eventtype       startdatetime  \
0       696255                Shooting Permit 2023-02-17 09:00:00   
1       714139                Shooting Permit 2023-05-12 13:00:00   
2       705334                Shooting Permit 2023-04-10 09:00:00   
3       705334                Shooting Permit 2023-04-10 09:00:00   
4       705334                Shooting Permit 2023-04-10 09:00:00   
...        ...                            ...                 ...   
12155   793926                Shooting Permit 2024-06-20 09:00:00   
12156   793926                Shooting Permit 2024-06-20 09:00:00   
12157   793783  Theater Load in and Load Outs 2024-06-20 00:01:00   
12158   785081  Theater Load in and Load Outs 2024-06-20 00:01:00   
12159   792643                Shooting Permit 2024-06-20 07:00:00   

              enddatetime           enteredon  \
0     2023-02-18 12:00:00 2023-02-14 22:47:33   
1     2023-05-13 05:00:00 2023-05-04 14:27:51   
2     2023-04-10 22:00:00 2023-03-30 17:17:29   
3     2023-04-10 22:00:00 2023-03-30 17:17:29   
4     2023-04-10 22:00:00 2023-03-30 17:17:29   
...                   ...                 ...   
12155 2024-06-20 22:00:00 2024-06-17 22:24:41   
12156 2024-06-20 22:00:00 2024-06-17 22:24:41   
12157 2024-06-24 23:59:00 2024-06-17 13:21:55   
12158 2024-06-21 06:00:00 2024-05-16 12:54:50   
12159 2024-06-20 22:00:00 2024-06-12 17:10:02   

                                   eventagency  \
0      Mayor's Office of Media & Entertainment   
1      Mayor's Office of Media & Entertainment   
2      Mayor's Office of Media & Entertainment   
3      Mayor's Office of Media & Entertainment   
4      Mayor's Office of Media & Entertainment   
...                                        ...   
12155  Mayor's Office of Media & Entertainment   
12156  Mayor's Office of Media & Entertainment   
12157  Mayor's Office of Media & Entertainment   
12158  Mayor's Office of Media & Entertainment   
12159  Mayor's Office of Media & Entertainment   

                                             parkingheld    borough  \
0      KINGSLAND AVENUE between DEAD END and GREENPOI...   Brooklyn   
1      WEST   26 STREET between 12 AVENUE and 11 AVEN...  Manhattan   
2      SOUTH STREET between BROAD STREET and OLD SLIP...  Manhattan   
3      SOUTH STREET between BROAD STREET and OLD SLIP...  Manhattan   
4      SOUTH STREET between BROAD STREET and OLD SLIP...  Manhattan   
...                                                  ...        ...   
12155  SCHERMERHORN STREET between COURT STREET and B...   Brooklyn   
12156  SCHERMERHORN STREET between COURT STREET and B...   Brooklyn   
12157  WEST   35 STREET between 8 AVENUE and 9 AVENUE...  Manhattan   
12158  FROST STREET between DEBEVOISE AVENUE and MORG...   Brooklyn   
12159  COLUMBUS AVENUE between WEST   62 STREET and W...  Manhattan   

      communityboard(s) policeprecinct(s)           category  subcategoryname  \
0                     1                94         Television   Cable-episodic   
1                     4                10         Television  Episodic series   
2                  1, 3              1, 5               Film          Feature   
3                  1, 3              1, 5               Film          Feature   
4                  1, 3              1, 5               Film          Feature   
...                 ...               ...                ...              ...   
12155           2, 3, 4    79, 81, 83, 84  Still Photography   Not Applicable   
12156           2, 3, 4    79, 81, 83, 84  Still Photography   Not Applicable   
12157                 4                14            Theater          Theater   
12158                 1                94            Theater          Theater   
12159              4, 7            18, 20         Television  Episodic series   

                        country zipcode  
0      United States of America   11222  
1      United States of America   10001  
2      United States of America   10

Once the data frame is clean, please answer the following questions with code:
1. Which borough has the most _unique_ events for television, film and theater respectively?
2. Which zip code has the most _unique_ events for television, film and theater respectively?

In [59]:
tv_film_permits = film_permits[film_permits['category']=='Television']
film_film_permits = film_permits[film_permits['category']=='Film']
theater_film_permits = film_permits[film_permits['category']=='Theater']

borough_unique_tv_events = tv_film_permits.groupby('borough')['eventid'].nunique()
borough_unique_events

borough
Bronx             223
Brooklyn         1199
Manhattan        1310
Queens            874
Staten Island       5
Name: eventid, dtype: int64

In [60]:
borough_with_most_unique_tv_events = borough_unique_events.idxmax()
borough_with_most_unique_tv_events

'Manhattan'

In [61]:
borough_unique_film_events = film_film_permits.groupby('borough')['eventid'].nunique()
borough_unique_film_events

borough
Bronx             44
Brooklyn         360
Manhattan        416
Queens           122
Staten Island     28
Name: eventid, dtype: int64

In [62]:
borough_with_most_unique_film_events = borough_unique_film_events.idxmax()
borough_with_most_unique_film_events

'Manhattan'

In [66]:
borough_unique_theater_events = theater_film_permits.groupby('borough')['eventid'].nunique()
borough_unique_theater_events

borough
Bronx               3
Brooklyn          343
Manhattan        1190
Queens              0
Staten Island       0
Name: eventid, dtype: int64

In [65]:
borough_with_most_unique_theater_events = borough_unique_theater_events.idxmax()
borough_with_most_unique_theater_events

'Manhattan'

In [68]:
zipcode_unique_tv_events = tv_film_permits.groupby('zipcode')['eventid'].nunique()
zipcode_unique_tv_events
zipcode_with_most_unique_tv_events = zipcode_unique_tv_events.idxmax()
zipcode_with_most_unique_tv_events

'11222'

In [70]:
zipcode_unique_film_events = film_film_permits.groupby('zipcode')['eventid'].nunique()
zipcode_unique_film_events
zipcode_with_most_unique_film_events = zipcode_unique_film_events.idxmax()
zipcode_with_most_unique_film_events

'10002'

In [71]:
zipcode_unique_theater_events = theater_film_permits.groupby('zipcode')['eventid'].nunique()
zipcode_unique_theater_events
zipcode_with_most_unique_theater_events = zipcode_unique_theater_events.idxmax()
zipcode_with_most_unique_theater_events

'10019'

## Answer

### Student name:
Gladwin Osakwe

### Your answer:
1. The borough with the most unique TV events is ... Manhattan
2. The borough with the most unique film events is ... Manhattan
3. The borough with the most unique theater events is ... Manhattan

4. The zip code with the most unique TV events is ... 11222
5. The zip code with the most unique film events is ... 10002
6. The zip code with the most unique theater events is ... 10019